Youtube Shortcode: Updates regex

kraftbj opened the pull request:
Brings upstream changes from wp.com discussed in D17568-code and D17868-code.

#### Changes proposed in this Pull Request:

* Updates the regex used in the Youtube embed sniffing to allow for some cases previously missed.

#### Testing instructions:

* See D17568-code

#### Proposed changelog entry for your changes:

* Updates YouTube regex for converting `object` era-embeds to a shortcode.

kraftbj committed:
Youtube Shortcode: Updates regex

Brings upstream changes from wp.com discussed in D17568-code and D17868-code.

kraftbj commented:
Convo in p1537995658000100-jetpack-crew , but before merging, we should run plenty of performance testing.

georgestephanis commented on modules/shortcodes/youtube.php:
```diff
@@ -55,10 +55,10 @@ function youtube_embed_to_short_code( $content ) {
 				$params = $match[1];
 
-				if ( 'ifr_regexp_ent' == $reg ) {
+				if ( in_array( $reg, array( 'ifr_regexp_ent', 'regexp_ent' ) ) ) {
```
Would it be simpler for this to do a `wp_endswith( $reg, '_ent' )` ?

westi commented:
There is a new version of code in D21652-code once that is reviewed we should sync this PR with the new code.

westi committed:
Improve the matching of the youtube shortcode reversal again.

Merges r184358-wpcom

In D17568-code and D17868-code we improved this code.
Those changes accidentally got reversed in D18511-code
I was reviewing #10224 which would merge the original
changes into Jetpack and noted the 's' modifier.
Unfortunately we can't include that here because it means on long but malformed post content we go into an
infinite loop because of the .*? in the RegEx.
This means we can't match objects/mangled objects which have a newline within them but we can match more
that we were

Diff: D21652-code

jeherve commented:
> once that is reviewed we should sync this PR with the new code.

Took care of that in e4c966d6ce59ef42260a4283d30af470229e2da5

pablinos commented:
You're right @andfinally, they should be escaped. It does currently match:

```
<object data="https://www youtubescom/v/caWbt0IOOZk" type="application/x-shockwave-flash" width="400" height="300"><param name="movie" value="https://www.youtube.com/v/caWbt0IOOZk" /><param name="quality" value="high" /><param name="allowFullScreen" value="true" /><!-- Fallback content --><a href="https://www.youtube.com/watch?v=caWbt0IOOZk"><img height="300" src="https://img.youtube.com/vi/caWbt0IOOZk/0.jpg" width="400" />YouTube Video</a></object>
```

Although the resulting shortcode would correct this as it only uses the video ID, it's definitely worth fixing.

andfinally committed:
YouTube: escape . characters when matching Youtube.com

@see https://github.com/Automattic/jetpack/pull/10224#commitcomment-31535800

jeherve commented:
👍 	Made the change in d3bd4ef82f5b9e2881f962082efe6c225c86e297

jeherve reviewed (approved):
🚢 	Merging this.
